### PR TITLE
fix: Configure table data rows count

### LINKED ISSUE #1578: [Feature] Enable num rows for server side pagination
## Summary

Tables with serverside pagination can be navigated infinitely. Having an option to specify the number of rows will allow us to determine the number of pages

### LINKED ISSUE #2230: [Feature] Allow users to configure a page size for server side pagination
## Motivation

This will allow us to disable the next page button if the page size is lower than the expected page size. Read more about the conversation here #2107 & #1578

### LINKED ISSUE #4614: [Bug] Next/previous is clickable on Table pagination after enabling server side pagination from property pane
### Steps to reproduce the behaviour:

1. Drag and drop table widget
2. Query a table data with multiple pages 
3. From property pane enable server side pagination
4. Observe that Next/Previous page is still clickable on table

### Important Details
 - Environment : Prod
 - OS: MacOSX
 - Browser chrome

[LOOM DEMO](https://www.loom.com/share/89d04b26cae7484fa0870ce5cb995007)

### LINKED ISSUE #5447: [Bug]The "Total record count" and "Page size" should only be visible when the server side pagination is enabled. Also add a tooltip/Helper function to the "Total record count" and "Page size".
## Description

[What happened]The "Total record count" and "Page size" should only be visible when the server side pagination is enabled. They should also have helpers to convey the functionality to the user.

### Steps to reproduce the behaviour:

1. Drag and drop the tablet widget on the canvas
2. when the server side pagination toggle is disabled
3. check if the "Total record count" and "Page size" input boxes are displayed

expected result : The "Total record count" and "Page size" input boxes should not be displayed when the server side pagination is disabled. Also add a tooltip/Helper function to the "Total record count" and "Page size".

### Important Details

- Version: Cloud
- OS: Windows
- Browser Chrome
- Environment Release

### LINKED ISSUE #5460: [Bug]The page size should change based on the table height and the default page size field should be removed.
## Description

[What happened]The page size should change based on the table height and the default page size field should be removed.

### Steps to reproduce the behaviour:

1. Drag and drop the table widget
2. Enable server side pagination
3. Resize the Table widget 
4. Observe the display of data

Expected result: The page size should change based on the table height and the data should change accordingly this should work for both server-side pagination and for the default pagination.

### Important Details

- Version: [Cloud ]
- OS: Windows
- Browser Chrome
- Environment Release

### LINKED ISSUE #5645: [Bug][deploy previe-5552] All the data is displayed in the table even the "Total record count " is specified
## Description

[What happened]All the data is displayed in the table even the "Total record count " is specified

### Steps to reproduce the behaviour:

1.Drag and drop table widget
2.Query a table data with multiple pages
3.From property pane enable server-side pagination
4.enter the value in total record count input box

Actual result: All the records present in the api is displayed in the table instead of 5 records

Expected result : The data should be displayed in table based on the total count added in the Total record count input box 

### Important Details

- Version: [Cloud ]
- OS: Windows
- BrowserChrome
- Environment [ deploy preview]

### PR DESCRIPTION
**Pull request by vicky-primathon**
## Description
Allow total records count to be configured for Table widget when server side pagination is enabled

Fixes #2230
Fixes #1578 
Fixes #4614
Fixes #5447 
Fixes #5645 
Fixes #5460 

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Added cypress test case to validate total records count property works 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: configure-table-data-rows-count 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.84 **(-0.01)** | 36.76 **(-0.03)** | 33.56 **(0.01)** | 55.42 **(-0.02)**
 :green_circle: | app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx | 61.63 **(0)** | 38.18 **(0.93)** | 38.89 **(0)** | 60.98 **(0)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/TableHeader.tsx | 50 **(0)** | 52 **(-2.17)** | 30 **(0)** | 47.37 **(0)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/index.tsx | 27.27 **(0)** | 2.56 **(-0.04)** | 15.79 **(0)** | 25.68 **(0)**
 :green_circle: | app/client/src/pages/Editor/PropertyPane/PropertyHelpLabel.tsx | 88.89 **(1.39)** | 87.5 **(0)** | 100 **(0)** | 88.89 **(1.39)**
 :red_circle: | app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts | 28.69 **(-1.13)** | 7.37 **(-1.06)** | 47.62 **(0.25)** | 29.91 **(-1.28)**
 :red_circle: | app/client/src/widgets/TableWidget/index.tsx | 27.41 **(-0.3)** | 6.08 **(-0.17)** | 25.49 **(0)** | 27.57 **(-0.33)**</details>

### CODE REVIEW

**Review by somangshu — DISMISSED** [review]
Approving to run the tests

**Comment by somangshu** [comment]
@RakshaKShetty Can you test this and the other issues raised around this. Please mention them here if possible

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx [hunk 1]
```diff
@@ -155,7 +159,7 @@ export function Table(props: TableProps) {
   }
   let startIndex = currentPageIndex * props.pageSize;
   let endIndex = startIndex + props.pageSize;
-  if (props.serverSidePaginationEnabled) {
+  if (props.serverSidePaginationEnabled && props.totalRecordsCount) {
```
[the remark this answers is not in the document]
Comment (vicky-primathon) [reply app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:162]: @SatishGandham  this is newly coded. Added check for totalRecordsCount as the data should be paginated if server side pagination is enabled and total records count is configured.

**Review by SatishGandham — DISMISSED** [review]

**Comment by RakshaKShetty** [comment]
@somangshu and @vicky-primathon there are 3 open issues. #5645
#4614
#5447

**Comment by somangshu** [comment]
@vicky-primathon we need to have the above fixed in this PR. Kindly also link them. 

@RakshaKShetty have you verified that the above mentioned exist in the Deploy preview?

**Comment by RakshaKShetty** [comment]
> @vicky-primathon we need to have the above fixed in this PR. Kindly also link them.
> 
> @RakshaKShetty have you verified that the above mentioned exist in the Deploy preview?

Ya @somangshu i have tested this in deploy preview

**Review by somangshu — DISMISSED** [review]

**Comment by somangshu** [comment]
@RakshaKShetty can you reverify the issues and give us a sign off here?

**Comment by RakshaKShetty** [comment]
1.The tooltip for Table "Total Record Count" is too big and goes outside the page creating the scroll bar
2.the total record count does not reset when the server-side pagination is disabled. 
3.The table should also refresh. when the server-side pagination is disabled
4.When server-side pagination is enabled and the user goes to the last page and resize the table, the last page become completely empty and we will have a lot of blank pages with no data since the page size is changed
screenshot is for point number 4
image

File: app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts [hunk 2]
```diff
@@ -682,7 +682,7 @@ export default [
       },
       {
         helpText:
-          "Bind the Table.pageSize property in your API. Without page size configured in your API, this functionality will not work as it uses page size to calculate total pages and whether next page button is disabled.",
+          "Bind the Table.pageSize and Table.pageNo property in your API and call it onPageChange. Without page size configured in your API, this functionality will not work as it uses page size to calculate total pages and whether next page button is disabled.",
```
Comment (somangshu) [inline app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts:685]: Change to > "Bind the Table.pageSize and Table.pageNo property in your API and call it onPageChange. Without this the Table widget cannot calculate the number of pages and disable page buttons",

File: app/client/src/widgets/TableWidget/index.tsx [hunk 3]
```diff
@@ -563,6 +579,8 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
     }
 
     if (this.props.pageSize !== prevProps.pageSize) {
+      //reset current page number when page size changes
```
Comment (somangshu) [inline app/client/src/widgets/TableWidget/index.tsx:582]: So if the page size is updated does the table data also get reset?

File: app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts (hunk 2 above)
Comment (vicky-primathon) [reply app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts:685]: sure

File: app/client/src/widgets/TableWidget/index.tsx (hunk 3 above)
Comment (vicky-primathon) [reply app/client/src/widgets/TableWidget/index.tsx:582]: Yes, when pageSize is used in query and we reset pageNo or when pageSize change is configured

**Review by somangshu — APPROVED** [review]
Looks good.

@RakshaKShetty please test the helper text tooltip properly across widgets.
Also kindly verify the above issues marked

**Comment by somangshu** [comment]
@Nikhil-Nandagopal can you check this? We have made changes based on the last conversation.

**Review by somangshu — APPROVED** [review]

**Comment by RakshaKShetty** [comment]
@vicky-primathon  still there is one issue that persists 
1.when the server-side pagination is enabled and the user resize the table then the data is not appended based on the table resize.
image

**Comment by vicky-primathon** [comment]
> @vicky-primathon still there is one issue that persists
> 1.when the server-side pagination is enabled and the user resize the table then the data is not appended based on the table resize.
> image

@RakshaKShetty we also need to configure API call on onPageSizeChange so that data is refreshed when user resize table.

**Comment by Nikhil-Nandagopal** [comment]
@somangshu 
the next button seems disabled regardless of the number of rows
the total record count should not limit the number of records on the table

**Comment by somangshu** [comment]
@Nikhil-Nandagopal thanks for pointing the obvious here. 

@vicky-primathon we need to make sure that when the `total record count` is not set the page disabling functionality does not kick into action.

@RakshaKShetty the page size refresh work just fine if we bind the same api to on page size changed. My bad, I did not explain this well. If you find any other issues kindly let us know.

**Review by SatishGandham — APPROVED** [review]

**Comment by vicky-primathon** [comment]
> @Nikhil-Nandagopal thanks for pointing the obvious here.
> 
> @vicky-primathon we need to make sure that when the `total record count` is not set the page disabling functionality does not kick into action.
> 
> @RakshaKShetty the page size refresh work just fine if we bind the same api to on page size changed. My bad, I did not explain this well. If you find any other issues kindly let us know.

@somangshu  fix this issues
1. Page disabling functionality should not start until total record count is configured.
2. total record count should not limit the number of records on the table.

**Review by SatishGandham — APPROVED** [review]

**Comment by somangshu** [comment]
@RakshaKShetty the issues are fixed here, Can we reverify these at the soonest.

There are the bugs pointed last:
> 1. Page disabling functionality should not start until total record count is configured.
> 2. total record count should not limit the number of records on the table.
> 3. all other related issues

**Comment by vivekverma2312** [comment]
@vicky-primathon 
1. Able to change page number beyond the last record 
https://www.loom.com/share/63a32316a71b4ecb89edeeb32d81846b

2. New records are not loading on scrolling the data in table (Before data use to load on scroll) 
3. Page size property is not showing on enabling server side pagination

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Review by SatishGandham — DISMISSED** [review]

**Review by somangshu — DISMISSED** [review]

**Comment by vivekverma2312** [comment]
@vicky-primathon Observations:
1. Error when total record property is not set
2. When total record count property is set then records are coming only on page 1 (Number of record  = page size)
3. New records are not loading on scrolling the records on the table for server side pagination.
https://www.loom.com/share/d40ed5d53b9b46b6bd660d2253637d04

**Comment by vivekverma2312** [comment]
@vicky-primathon Only issue 2 is resolved in the latest deploy preview. Issue 1 and 3 still exist.
4. Changing the table height changes the page no to page1
https://www.loom.com/share/cc9803888aa544fa8a006eda27866cb3

**Review by somangshu — APPROVED** [review]

**Review by SatishGandham — APPROVED** [review]

**Pull request merged**